libgit/autogit_manager.go

```diff
@@ -0,0 +1,397 @@
+	defer func() {
+		am.deferLog.CDebugf(ctx, "Reset request completed: %+v", err)
+	}()
+
+	// Make a new single-op config for processing this request.
+	ctx, gitConfig, tempDir, err := am.getNewConfig(ctx)
+	if err != nil {
+		return err
+	}
+	// Cloned/pull data should be regular data blocks.
+	gitConfig.SetDefaultBlockType(keybase1.BlockType_DATA)
+	defer func() {
+		gitConfig.Shutdown(ctx)
+		rmErr := os.RemoveAll(tempDir)
+		if rmErr != nil {
+			am.log.CDebugf(
+				ctx, "Error cleaning storage dir %s: %+v\n", tempDir, rmErr)
+		}
+	}()
+
+	uniqID, err := makeUniqueID(ctx, gitConfig)
+	if err != nil {
+		return err
+	}
+
+	// Construct a src repo FS.
+	srcRepoFS, _, err := GetRepoAndID(
+		ctx, gitConfig, req.srcTLF, req.srcRepo, uniqID)
+	if err != nil {
+		return err
+	}
+
+	// And a dst parent checkout FS, which better already exist.
+	dstFS, err := libfs.NewFS(
+		ctx, gitConfig, req.dstTLF, req.dstDir, uniqID,
+		keybase1.MDPriorityNormal)
+	if err != nil {
+		return err
+	}
+
+	// Take the lock for the dst repo while resetting.
+	lockFile, err := dstFS.Create(autogitLockName(req.srcRepo))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Because we took the lock, this Close will sync/flush the
+		// whole journal.
+		closeErr := lockFile.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	err = lockFile.Lock()
+	if err != nil {
+		return err
+	}
+
+	dstRepoFS, err := dstFS.Chroot(req.srcRepo)
+	if err != nil {
+		return err
+	}
+
+	// For now, assume the branch name refers to a ref head.
+	branch := plumbing.ReferenceName(
+		fmt.Sprintf("refs/heads/%s", req.branchName))
+	am.log.CDebugf(ctx, "Starting the reset")
+	return Reset(ctx, srcRepoFS, dstRepoFS, branch)
+}
+
+func (am *AutogitManager) clearFromQueue(req resetReq) (
+	waitCh <-chan struct{}) {
+	am.lock.Lock()
+	defer am.lock.Unlock()
+	inProgress, ok := am.resetsInProgress[req.dstPath()]
+	if ok {
+		return inProgress.doneCh
+	}
+	delete(am.resetsInQueue, req.dstPath())
+	am.resetsInProgress[req.dstPath()] = req
+	return nil
+}
+
+func (am *AutogitManager) clearFromInProgress(req resetReq) {
+	am.lock.Lock()
+	defer am.lock.Unlock()
+	delete(am.resetsInProgress, req.dstPath())
+}
+
+func (am *AutogitManager) resetWorker(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for req := range am.resetQueue {
+		ctx := libkbfs.BackgroundContextWithCancellationDelayer()
+		ctx = libkbfs.CtxWithRandomIDReplayable(
+			ctx, ctxIDKey, ctxOpID, am.log)
+		for {
+			waitCh := am.clearFromQueue(req)
+			if waitCh == nil {
+				break
+			}
+			// We need to wait for any in-progress resets, since the
+			// File.Lock is taken per-instance, not per-goroutine.
+			am.log.CDebugf(ctx, "Waiting to process reset request from "+
+				"%s/%s to %s/%s",
+				req.srcTLF.GetCanonicalPath(), req.srcRepo,
+				req.dstTLF.GetCanonicalPath(), req.dstDir)
+			<-waitCh
+			am.log.CDebugf(ctx, "Done waiting")
+		}
+
+		_ = am.doReset(ctx, req)
+		am.clearFromInProgress(req)
+		close(req.doneCh)
```

Comment:
> Actually thinking about it a bit more, maybe you can close the channel first. Then if another worker races to get the in-progress request, it'll get the already-closed channel, rather than creating a new request.

A race on the in-progress channel doesn't result in a new request.  It only applies to the `for` loop right above here, which waits for any in-progress request to finish before starting the next one.  (The next one has to be done, because it must have been started after the current reset started, and therefore might refer to a new head.)  You might be confusing it with the entry in `resetsInProgress`, which has already been cleared above.

I think it's somewhat irrelevant whether it's cleared from in-progress first or closed first.  I'll clarify in a comment.